### DISCUSSION TITLE
openapi.yamlをどの場所に置くかについて

### DISCUSSION THREAD

**Discussion by s1-haya** [42-pong/42-pong#36] [Ideas]
### 背景
https://github.com/42-pong/42-pong/pull/27#pullrequestreview-2425609291

**Comment by habvi** [comment]
そもそもあまり frontend, backend が分かれてて Django で…という例を見つけるのが難しいのですが、見つけた中ではルートに専用のリポジトリ `openapi/` , `api/` を置いてその中に入れているのがありました
- `frontend/`, `backend/` が分かれているリポジトリの例
  https://github.com/PostHog/posthog/tree/master/openapi
- 分かれてないが有名どころ
  https://github.com/kubernetes/kubernetes/tree/master/api

ただ swagger-ui について調べていて知ったのですが、swagger-ui 用に入れた Django のサードパーティパッケージ `drf-spectacular` が自動で OpenAPI スキーマを生成してくれて、`@extend_schema` でスキーマのカスタマイズを view に書いて使えたりするので、
https://drf-spectacular.readthedocs.io/en/latest/readme.html#usage
この方法を frontend との API にも採用した場合は、YAML 形式で保存しておくことがなくなるかもしれないです

YAML 形式か、自動生成か、など含め相談して決めていきたいです🙏

**Reply by ayase-mstk** [reply]
自動でOpenAPIスキーマを生成するということは、実装してあるものを基に生成されることになるので、設計を先にしないということになるような気がします。

自動で生成されると常に最新の状態がキープされ、それもメリットではあるので、ある程度APIの実装が進んでいって細かい修正などが入るようになったタイミングで自動生成に切り替えるというのも一つの手かなと思いました。

**Reply by habvi** [reply]
まだ API の全容を分かってなくて勘違いしてるかもですみませんなのですが、自動で OpenAPI スキーマを作成するイコール実装が終わってる必要がある、ということでしょうか？

調べていて以下のような流れをイメージしていたのですが、他にもあれば知りたいです🙌
- endpoint の仕様が決まったらコード内に YAML の中身とほぼ同じような内容 (`@extend_schema`) と、urlpatterns だけを作る
- endpoint が呼ばれた時の実際の実装 `add_user()` 自体はまだの状態で、swagger-ui で frontend と仕様を共有
- backend も仕様を満たす実装を `add_user()` に追加していく


user/views.py ?
```py
@extend_schema(
    summary="Add a new user",
    description="This endpoint allows you to add a new user to the system.",
    responses={
        200: OpenApiResponse(
            response={"message": "User added successfully"},
            description="A successful response indicating the user was added.",
        )
    },
    examples=[
        OpenApiExample(
            "Example 1",
            value={"message": "User added successfully"},
            description="An example of a successful response.",
        )
    ],
)
@api_view(["POST"])
def add_user(request):
    # TODO: 実装をする
    return Response({"message": "todo"})
```
user/urls.py
```py
urlpatterns = [
    path("add/", add_user, name="add_user"),
]
```
urls.py
```py
urlpatterns = [
    path("admin/", admin.site.urls),
    # swagger-ui
    path("api/schema/", SpectacularAPIView.as_view(), name="schema"),
    path(
        "api/schema/swagger-ui/",
        SpectacularSwaggerView.as_view(url_name="schema"),
        name="swagger-ui",
    ),
    # app
    path("api/user/", include("user.urls")),
]
```

**Reply by ayase-mstk** [reply]
完全に誤解してました。modelsとかも作らないといけないかと勘違いしていました。
細かい内容も`@extend_schema_field`や`@extend_schema_serializer`などで追加できるのですね。

確かにそれができるのであれば、直接openapi.ymlを書く必要はないですね。
@sawamura-hayato さんごめんなさい。。

**Reply by habvi** [reply]
私も合ってるか自信ないのですが…、早めに気付けて良かったです！
どっちの方針でいくかはまだ分からなさそうですね

openapi.yml も書いてる内容はほぼ同じだと思うので、調べていただいたのはそのまま使える気がします🙏

**Comment by s1-haya** [comment]
ご意見ありがとうございます。とても参考になりました。僕的には初め自動生成にして、後から細かい設定が必要な場合は手動に切り替えてもいいのなって思いました。

自動生成
- コードに基づいて自動で更新してくれる
- openapi.yamlの位置を決めなくて済む

手動
- 細かい設定ができる

@ayase-mstk  謝らないでください！OpenAPIについてとても勉強になりました！
